packages/caravan-fees/src/cpfp.ts

```diff
@@ -0,0 +1,301 @@
+    );
+  }
+
+  childTxTemplate.addInput(BtcTxInputTemplate.fromUTXO(parentOutputUTXO));
+
+  // Step 5: Add a change output (at least 1 output required for a valid transaction)
+  childTxTemplate.addOutput(
+    new BtcTxOutputTemplate({
+      address: changeAddress,
+      amountSats: "0", // Initial amount, will be adjusted later
+    }),
+  );
+
+  /**
+   * Step 6: Add additional inputs to cover the CPFP fee requirements
+   *
+   * This step implements a simplified version of the "Child-pays-for-parent" (CPFP)
+   * transaction selection logic, inspired by Bitcoin Core's implementation
+   * (https://github.com/bitcoin/bitcoin/pull/7600).
+   *
+   * In Bitcoin Core, transactions are sorted in the mempool using a modified
+   * feerate that considers ancestors. During block creation, transactions are
+   * considered in this order. As transactions are selected, a separate map tracks
+   * the new feerate-with-ancestors for in-mempool descendants.
+   *
+   * Our simplified approach:
+   * 1. We start with the parent transaction's fee and vsize.
+   * 2. We iteratively add inputs to the child transaction.
+   * 3. After each input addition, we recalculate the combined fee rate of the
+   *    parent and child transactions.
+   * 4. We continue adding inputs until the combined fee rate meets or exceeds
+   *    the target fee rate.
+   *
+   * This method ensures that the child transaction provides enough fee to incentivize
+   * miners to include both the parent and child transactions in a block, effectively
+   * "bumping" the fee of the parent transaction.
+   *
+   * Note: While this approach is simpler than Bitcoin Core's full implementation,
+   * it achieves the core goal of CPFP: ensuring the combined package (parent + child)
+   * is attractive for miners to include in a block.
+   */
+
+  const parentFee = new BigNumber(analysis.fee);
+  const parentVsize = analysis.vsize;
+
+  for (const utxo of availableInputs) {
+    if (
+      isCPFPFeeSatisfied(
+        parentFee,
+        parentVsize,
+        new BigNumber(childTxTemplate.currentFee),
+        childTxTemplate.estimatedVsize,
+        parseFloat(txAnalyzer.cpfpFeeRate),
+      )
+    ) {
+      break; // Stop adding inputs once CPFP fee requirements are met
+    }
+    // Skip if this UTXO is already added
+    if (
+      childTxTemplate.inputs.some(
+        (input) => input.txid === utxo.txid && input.vout === utxo.vout,
+      )
+    ) {
+      continue;
+    }
+    childTxTemplate.addInput(BtcTxInputTemplate.fromUTXO(utxo));
+  }
+
+  // Step 7: Calculate and set the change output amount
+  const totalInputAmount = new BigNumber(childTxTemplate.getTotalInputAmount());
+  const fee = new BigNumber(childTxTemplate.targetFeesToPay);
+  const changeAmount = totalInputAmount.minus(fee);
+
+  if (changeAmount.lte(dustThreshold)) {
+    if (strict) {
+      throw new Error(
```

Comment:
It concerns me this check is happening here. Shouldn't we be accounting for this when we're adding inputs? Is this just in the case there were not enough inputs? 

My guess is probably, but then this code is confusing. The error should probably read that more inputs are needed ("increase inputs" was a little confusing) and then if this is the only scenario where we'd get here, what would `adjustChangeOutput` be able do for a CPFP with just one output?

Reply:
Ahh yes you are right !!  `adjustChangeOutput` won't help in this case , so we have a single output in this case 
```ts 
  childTxTemplate.addOutput(
    new BtcTxOutputTemplate({
      address: changeAddress,
      amountSats: "0", // Initial amount, will be adjusted later
    }),
  );
``` 
and by default all outputs are malleable ,unless locked on initialisation ....

So what we are doing is that we have added the required inputs and we have a total amount (from that inputs ) and we now know the fees we have to pay ... so we set the output amount to be just `inputs - fee `

so that's why the check 
```ts
if (changeAmount.lte(dustThreshold)) {
    if (strict) {
      throw new Error(
        "Change amount is below the dust threshold. Increase inputs or reduce fee rate.",
      );
```
tells that we don't have enough amount of inputs to cover the fees for both `parent + child ` , in strict mode it throws an error and else the correct code should look like ...

```ts

  if (changeAmount.lte(dustThreshold)) {
    if (strict) {
      throw new Error(
        "Change amount is below the dust threshold. More inputs are needed or try reducing the fee rate.",
      );
    } else {
      console.warn(
        "Change amount is below the dust threshold. More inputs are needed or try reducing the fee rate",
      );

// we'll use this instead of adjustChangeOutput fn
     childTxTemplate.outputs[0].setAmount(changeAmount.toString());
    }
  } else {
    childTxTemplate.outputs[0].setAmount(changeAmount.toString());
  }
  ```

Reply:
There’s some duplication here you could probably clean up but I think that looks right!

Reply:
Also I guess 
```ts
 childTxTemplate.outputs[0].setAmount(changeAmount.toString());
 ```
 
 and 
 
 ```ts
 childTxTemplate.adjustChangeOutput();
 ```
 
 would do the same thing ... with ` childTxTemplate.adjustChangeOutput()` providing extra validation

Reply:
changes [done](https://github.com/caravan-bitcoin/caravan/pull/114/commits/1109c68ea23fa1a25e9acea12533fb78ab1bed84#diff-852fd83c241ed78594dc29389ec04d1425d784e665c706434a8de046f7daa85dR198)